src/types/Writable.ts

```diff
@@ -15,6 +15,12 @@ type WritableKeys<T> = {
 	[P in keyof T]: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, P>;
 }[keyof T];
 
+type OptionalKeys<T> = {
+	[P in keyof T]: {} extends Pick<T, P> ? P : never;
+}[keyof T];
```

Comment:
What does `[keyof T]` do here? If I delete it, the TypeScript compiler and ESLint are both still happy. Is that because we're not using some edge case this addresses?

Reply:
They're happy, but it changes what `OptionalKeys` actually accomplishes

My understanding is that the `{}` block here is creating a mapped object where the attributes are keys of T, and values are either P or never.  Then the `[keyof T]` is going through each key and for keeping anything that isn't `never`.

With the `[keyof T]`:

<img width="479" alt="image" src="https://github.com/user-attachments/assets/fe2b33dd-2c84-4cdb-a70c-09b5bd6e6bac" />


Without `[keyof T]`:
<img width="479" alt="image" src="https://github.com/user-attachments/assets/dc7f86f6-c9aa-4081-817d-0fae52ac6777" />

So, without keeps the non-optional keys in there (but calls it `never`) <-- the presence of the key in the object means it gets kept later on.

Reply:
Aha! Excellent, thank you very much for the explanation and demonstration!